File: src/lib/text-select.ts
/**
 * I-beam text-selection tool: IPC wrappers + pure gesture math. A dedicated file
 * (mirrors the recent-docs.ts pattern) so the text-selection feature does not
 * touch ipc.ts's command-wrapper section beyond the one-line MarkupGeometry
 * addition - keeps the conflict surface small while another branch works on
 * the open/save/password path.
 *
 * PDFium is !Send + !Sync and lives on redline's dedicated render thread; these
 * commands are thin passthroughs (see src-tauri/src/commands/text_select.rs).
 */
import { invoke } from "@tauri-apps/api/core";
import type { PdfPoint } from "./ipc";

// ---------------------------------------------------------------------------
// Types (mirror Rust geometry::Quad / text::TextRangeSelection)
// ---------------------------------------------------------------------------

/**
 * A single quadrilateral in PDF user space, 4 points ordered top-left,
 * top-right, bottom-left, bottom-right. Mirrors Rust `geometry::Quad` - see
 * that type's doc comment for why this order (the de-facto Acrobat/Bluebeam
 * `/QuadPoints` convention, not the PDF spec's literal wording).
 */
export type Quad = [PdfPoint, PdfPoint, PdfPoint, PdfPoint];

/** Mirrors Rust `text::TextRangeSelection`. */
export interface TextRangeSelection {
  /** One quad per visual text line in the range (never merged). */
  quads: Quad[];
  /** Plain-text content of the range, for clipboard copy. */
  text: string;
}

// ---------------------------------------------------------------------------
// IPC wrappers
// ---------------------------------------------------------------------------

/**
 * Hit-test a PDF-user-space point to the nearest character index on a page,
 * within `tolerance` PDF points. Returns null when no character is within
 * tolerance (whitespace, image-only region, or off the text layer entirely).
 */
export async function charIndexAtPoint(
  docId: string,
  pageIndex: number,
  x: number,
  y: number,
  tolerance: number,
): Promise<number | null> {
  return invoke<number | null>("char_index_at_point", { docId, pageIndex, x, y, tolerance });
}

/**
 * Resolve a character range `[start, end)` on a page into the quads for a
 * text-anchored Highlight annotation plus the plain-text content for the
 * clipboard. Returns an empty selection (quads: [], text: "") for a degenerate
 * range - safe to call speculatively during a drag.
 */
export async function getTextSelection(
  docId: string,
  pageIndex: number,
  start: number,
  end: number,
): Promise<TextRangeSelection> {
  return invoke<TextRangeSelection>("get_text_selection", { docId, pageIndex, start, end });
}

// ---------------------------------------------------------------------------
// Pure gesture math (no DOM, no Tauri - unit-testable in isolation)
// ---------------------------------------------------------------------------

/**
 * Turn a drag's anchor (pointer-down hit) and focus (current pointer position
 * hit) character indices into a normalized `[start, end)` character range,
 * inclusive of BOTH the anchor and focus characters. Handles a backwards drag
 * (focus before anchor) by swapping. A single-character selection (anchor ===
 * focus) yields a 1-character range, not an empty one.
 */
export function selectionRange(anchorChar: number, focusChar: number): { start: number; end: number } {
  const lo = Math.min(anchorChar, focusChar);
  const hi = Math.max(anchorChar, focusChar);
  return { start: lo, end: hi + 1 };
}

File: src-tauri/src/commands/text_select.rs
//! Tauri commands for the I-beam text-selection tool (redline text-selection +
//! text-anchored highlight feature).
//!
//! Deliberately its own file (not `commands/text.rs`, which owns in-document
//! search/M4 S3) so it can be edited without touching that file - both commands
//! here are pure passthroughs to `RenderHandle`, which does the actual PDFium
//! work on the dedicated render thread.

use tauri::State;

use crate::text::TextRangeSelection;
use crate::AppState;

/// Hit-test a PDF-user-space point (`x`, `y`, same coordinate system as markups -
/// spec §5 invariant) to the nearest character index on `page_index`, within
/// `tolerance` PDF points. Returns `None` when no character is within tolerance.
///
/// Called on pointer-down and pointer-move while the I-beam tool is dragging, to
/// resolve the drag anchor/focus into character indices.
#[tauri::command]
pub async fn char_index_at_point(
    state: State<'_, AppState>,
    doc_id: String,
    page_index: u32,
    x: f64,
    y: f64,
    tolerance: f64,
) -> Result<Option<usize>, String> {
    state
        .render
        .char_index_at_point(doc_id, page_index, x, y, tolerance)
        .await
        .map_err(|e| format!("{e:#}"))
}

/// Resolve a character range `[start, end)` on `page_index` into the `Quad`s for
/// a text-anchored Highlight annotation plus the plain-text content for the
/// clipboard. Returns an empty selection (not an error) for `end <= start`.
#[tauri::command]
pub async fn get_text_selection(
    state: State<'_, AppState>,
    doc_id: String,
    page_index: u32,
    start: usize,
    end: usize,
) -> Result<TextRangeSelection, String> {
    state
        .render
        .text_range_selection(doc_id, page_index, start, end)
        .await
        .map_err(|e| format!("{e:#}"))
}
